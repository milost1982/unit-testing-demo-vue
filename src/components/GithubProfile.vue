<template>
  <div class="github-profile-component">
    <div v-if="profile">
      <img alt="Avatar"
           height="200"
           width="200"
           class="avatar"
           :src="profile.avatar_url"
      >
      <div class="description">
        <span class="profile-name">{{ profile.name }}</span>
        <span class="profile-login">Username: {{ profile.login }}</span>
        <span class="profile-company">Company: {{ profile.company }}</span>
        <span class="profile-location">Location: {{ profile.location }}</span>
        <span class="profile-blog">Blog: <a :href="profile.blog" target="_blank">{{ profile.blog }}</a></span>
        <span class="profile-url"><a :href="profile.html_url" target="_blank">Github Page</a></span>
      </div>
    </div>
    <div v-if="fetchError">
      <p style="color: red">{{ fetchError }}</p>
    </div>
  </div>
</template>

<script>
import githubLib from '@/lib/githubLib'

export default {
  data () {
    return {
      profile: null,
      fetchError: null
    }
  },
  // mounted () {
  //   this.getGithubProfile(this.username)
  // },
  methods: {
    async getGithubProfile (username) {
      try {
        this.profile = null
        this.fetchError = null

        this.profile = await githubLib.getGithubProfile(username)
      } catch (err) {
        console.error(err)
        this.fetchError = err
      }
    }
  }
}
</script>

<style scoped>
.avatar {
  height:auto;
  border-radius: 50%;
}
.description > span {
  display: block;
  margin: 0 0 8px 0;
}
.profile-name {
  font-size: 1.5em;
  font-weight: bold;
  margin: 5px 0;
}

</style>
