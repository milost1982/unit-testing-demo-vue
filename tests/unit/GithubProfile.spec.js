import { mount } from '@vue/test-utils'
import axios from 'axios'
import GithubProfile from '@/components/GithubProfile.vue'

describe('GithubProfile Component', () => {
  it('renders proper html from github profile data', async () => {
    // arrange
    const mockData = JSON.parse(`{
      "avatar_url": "https://avatars.githubusercontent.com/u/150330?v=4",
      "name": "Kyle Simpson",
      "login": "getify",
      "company": "Getify Solutions",
      "location": "Austin, TX",
      "html_url": "https://github.com/getify",
      "blog": "http://getify.me"
    }`)

    const originalAxiosGet = axios.get
    async function mockAxiosGet () {
      return { data: mockData }
    }
    axios.get = jest.fn(mockAxiosGet);

    const username = 'getify'
    const expectedUrl = `https://api.github.com/users/${username}`

    const wrapper = mount(GithubProfile)
    // await wrapper.find('input[type=submit]').trigger('submit')
    await wrapper.vm.getGithubProfile(username)

    // asserts

    // asserts for mocked function
    expect(axios.get).toHaveBeenCalledTimes(1)
    expect(axios.get).toHaveBeenCalledWith(expectedUrl)
    expect(wrapper.vm.$data.profile).toEqual(mockData)

    // asserts for rendered HTML
    // expect(wrapper.text()).toMatch(msg)
    expect(wrapper.find('.avatar').attributes().src).toBe(mockData.avatar_url)
    expect(wrapper.find('.profile-name').text()).toMatch(mockData.name)
    expect(wrapper.find('.profile-login').text()).toMatch(mockData.login)
    expect(wrapper.find('.profile-company').text()).toMatch(mockData.company)
    expect(wrapper.find('.profile-location').text()).toMatch(mockData.location)
    expect(wrapper.find('.profile-blog').text()).toMatch(mockData.blog)
    expect(wrapper.find('.profile-url > a').attributes().href).toBe(mockData.html_url)

    expect(wrapper).toMatchSnapshot()

    // cleanup
    axios.get = originalAxiosGet
  })
})
