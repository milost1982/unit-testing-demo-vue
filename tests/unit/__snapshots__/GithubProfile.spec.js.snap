// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubProfile Component renders proper html from github profile data 1`] = `
<div class="github-profile-component">
  <div><img alt="Avatar" height="200" width="200" src="https://avatars.githubusercontent.com/u/150330?v=4" class="avatar">
    <div class="description"><span class="profile-name">Kyle Simpson</span> <span class="profile-login">Username: getify</span> <span class="profile-company">Company: Getify Solutions</span> <span class="profile-location">Location: Austin, TX</span> <span class="profile-blog">Blog: <a href="http://getify.me" target="_blank">http://getify.me</a></span> <span class="profile-url"><a href="https://github.com/getify" target="_blank">Github Page</a></span></div>
  </div>
  <!---->
</div>
`;
